module/executiondatasync/optimistic_sync/execution_result/info_provider.go

```diff
@@ -202,75 +214,100 @@ func (e *Provider) validateCriteria(
-				receipts: executionReceiptList,
-			})
+	matchingResults := make([]resultWithReceipts, 0)
+	var lastErr error
+	for executionResultID, executionReceiptList := range allReceiptsForBlock.GroupByResultID() {
+		result := &executionReceiptList[0].ExecutionResult
+		receipts := executionReceiptList.GroupByExecutorID()
+
+		if err := p.isExecutorGroupMeetingCriteria(result, receipts, criteria); err != nil {
+			// skip groups that don't meet criteria; remember the last error
+			lastErr = err
+			continue
 		}
+
+		matchingResults = append(matchingResults, resultWithReceipts{
+			id:       executionResultID,
+			receipts: executionReceiptList,
+		})
 	}
 
-	if len(results) == 0 {
-		return nil, nil, common.NewInsufficientExecutionReceipts(blockID, 0)
+	if len(matchingResults) == 0 {
+		if lastErr != nil {
+			return flow.ZeroID, nil, lastErr
+		}
+		return flow.ZeroID, nil, optimistic_sync.ErrNotEnoughAgreeingExecutors
 	}
 
-	// sort results by the number of execution nodes in descending order
-	sort.Slice(results, func(i, j int) bool {
-		return len(results[i].receipts) > len(results[j].receipts)
+	// sort matchingResults by the number of execution nodes in descending order
+	sort.Slice(matchingResults, func(i, j int) bool {
+		return len(matchingResults[i].receipts) > len(matchingResults[j].receipts)
 	})
 
-	executorIDs := getExecutorIDs(results[0].receipts)
-	return results[0].result, executorIDs, nil
+	executorIDs := getExecutorIDs(matchingResults[0].receipts)
+	return matchingResults[0].id, executorIDs, nil
 }
 
 // isExecutorGroupMeetingCriteria checks if an executor group meets the specified criteria for execution receipts matching.
-func isExecutorGroupMeetingCriteria(
-	executorGroup flow.ExecutionReceiptGroupedList,
+//
+// Expected errors during normal operations:
+//   - [optimistic_sync.ErrForkAbandoned]: If the execution result is in a different fork than the one specified in the criteria.
+//   - [optimistic_sync.ErrNotEnoughAgreeingExecutors]: If the group does not have enough agreeing executors.
+//   - [optimistic_sync.ErrRequiredExecutorNotFound]: If the required executor is not in the group.
+func (p *Provider) isExecutorGroupMeetingCriteria(
+	executionResult *flow.ExecutionResult,
+	executorToExecutionReceipts flow.ExecutionReceiptGroupedList,
 	criteria optimistic_sync.Criteria,
-) bool {
-	if uint(len(executorGroup)) < criteria.AgreeingExecutorsCount {
-		return false
+) error {
+	if uint(len(executorToExecutionReceipts)) < criteria.AgreeingExecutorsCount {
+		return optimistic_sync.ErrNotEnoughAgreeingExecutors
 	}
 
+	// First, ensure the execution result is in the same execution fork, if any.
+	// This avoids returning other errors (like required executor missing) for a fork
+	// that we wouldn't consider anyway.
+	if criteria.ParentExecutionResultID != flow.ZeroID &&
+		executionResult.PreviousResultID != criteria.ParentExecutionResultID {
+		return optimistic_sync.ErrForkAbandoned
```

Comment:
this doesn't mean the fork was abandoned, just that the provided result's fork does not include the parent result.